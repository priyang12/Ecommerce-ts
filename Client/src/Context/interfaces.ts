import { AuthActions } from './AuthReducer';

export type AuthValues = {
  email: string;
  password: string;
};
export interface User {
  name: string;
  isAdmin: boolean;
}

export interface AuthState {
  loading: boolean;
  err: string | null;
  token: string | null;
  user: User | null;
}

export interface IContextModel {
  state: AuthState;
  dispatch: React.Dispatch<AuthActions>;
}
