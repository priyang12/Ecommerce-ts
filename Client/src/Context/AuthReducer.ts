import {
  LOGIN_SUCCESS,
  AUTH_ERROR,
  LOAD_USER,
  SET_LOADING,
  LOG_OUT,
} from './Authtypes';

import { AuthState, User } from './interfaces';

type AuthPayload = {
  [LOGIN_SUCCESS]: string | null;
  [AUTH_ERROR]: string | null;
  [LOAD_USER]: User | null;
  [SET_LOADING]: User | null;
  [LOG_OUT]: null;
};

type ActionMap<M extends { [index: string]: any }> = {
  [Key in keyof M]: M[Key] extends undefined
    ? {
        type: Key;
      }
    : {
        type: Key;
        payload: M[Key];
      };
};

export type AuthActions = ActionMap<AuthPayload>[keyof ActionMap<AuthPayload>];

export const AuthReducer = (state: AuthState, action: AuthActions) => {
  switch (action.type) {
    case LOGIN_SUCCESS:
      return {
        ...state,
        token: action.payload,
        loading: false,
      };
    case LOAD_USER:
      return {
        ...state,
        user: action.payload,
        loading: false,
      };
    case SET_LOADING:
      return {
        ...state,
        loading: true,
      };
    case AUTH_ERROR:
      return {
        ...state,
        loading: false,
        err: action.payload,
      };
    case LOG_OUT:
      return {
        ...state,
        token: null,
        loading: false,
        err: null,
      };
  }
};
