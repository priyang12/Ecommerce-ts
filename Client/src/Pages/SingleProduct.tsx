import { useState } from "react";
import { useParams } from "react-router-dom";
import { useFetch, useAxios } from "../Utils/CustomHooks";
import { DetailedProduct } from "../interfaces";
import Reviews from "../Components/Reviews";
import Rating from "../Components/Rating";
import Quantity from "../Components/Quantity";
import TimeoutBtn from "../Components/TimeoutBtn";
import AlertDisplay from "../Components/AlertDisplay";

//styles
import {
  StyledContainer,
  StyledProduct,
  StyledDetails,
  StyledCheckout,
} from "../Components/StyledComponents/Productdetails";

const SingleProduct = () => {
  const { id } = useParams<{ id: string }>();
  const [FetchData, ProductError, loading] = useFetch(
    `/api/products/product/${id}`
  );

  const [Qty, setQty] = useState("1");
  const [Params, setParams] = useState<any>({
    url: "",
  });
  const { Alert, Err } = useAxios(Params);

  const AddToCart = (e: React.ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    console.log("click");
    const CartItem = {
      id: Product._id,
      qty: Qty,
    };
    setParams({
      method: "POST",
      url: "/api/cart",
      data: CartItem,
    });
  };
  console.log(FetchData?.countInStock);
  if (loading) return <div data-testid='Loading'>Loading</div>;

  if (ProductError) return <div className='Error'>{ProductError}</div>;

  if (!FetchData) return null;

  const Product: DetailedProduct = FetchData;

  return (
    <div>
      {FetchData && (
        <StyledContainer>
          <StyledProduct>
            <img src={Product.image} alt={Product.name} />
            <StyledDetails>
              <h2>{Product.name}</h2>
              <div className='star-review'>
                <span className='stars'>
                  <Rating
                    value={Product.rating}
                    text={`${Product.numReviews} reviews`}
                  />
                </span>
              </div>
              <h4>Price: {Product.price}</h4>
              <div className='Description'>
                <div>Description:</div>
                {Product.description}
              </div>
            </StyledDetails>
            <StyledCheckout>
              <h3 className='status-label'>Price: {Product.price}</h3>
              {Product.countInStock <= 0 ? (
                <h3 style={{ color: "red" }}>Status: Out In Stock</h3>
              ) : (
                <h3 className='status-label'>Status: In Stock</h3>
              )}

              {Product.countInStock > 0 && (
                <form onSubmit={AddToCart}>
                  <label>Qty</label>
                  <select onChange={(e) => setQty(e.target.value)}>
                    <Quantity countInStock={Product.countInStock} />
                  </select>
                  <br />

                  {localStorage.token ? (
                    <TimeoutBtn
                      Time={5000}
                      className='btn'
                      FormValue='ADD TO CART'
                    />
                  ) : (
                    <div className='btn-gray'>Login/Register</div>
                  )}
                </form>
              )}

              {Alert && <AlertDisplay msg={Alert} type={true} />}
              {Err && <AlertDisplay msg={Err} type={false} />}
            </StyledCheckout>
          </StyledProduct>
        </StyledContainer>
      )}
      <Reviews reviews={FetchData?.reviews} />
    </div>
  );
};

export default SingleProduct;
