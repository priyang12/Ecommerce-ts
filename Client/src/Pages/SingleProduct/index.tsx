import { useState } from "react";
import { useParams } from "react-router-dom";
import { DetailedProduct } from "../../interfaces";
import Reviews from "../../Components/Reviews";
import Rating from "../../Components/Rating";
import Quantity from "../../Components/Quantity";
import TimeoutBtn from "../../Components/TimeoutBtn";
import AlertDisplay from "../../Components/AlertDisplay";

//styles
import {
  StyledContainer,
  StyledProduct,
  StyledDetails,
  StyledCheckout,
} from "../../Components/StyledComponents/Productdetails";
import axios from "axios";
import { useQuery, useMutation } from "react-query";
import { AddToCartQuery } from "../../API/CartAPI";
import { queryClient } from "../../query";
import Spinner from "../../Components/Spinner";
import { Helmet } from "react-helmet";
import { SingleProductCall } from "../../API/ProductAPI";
import { AddWishlistQuery } from "../../API/WishListAPI";

const SingleProduct = () => {
  const { id } = useParams<{ id: string }>();
  const [Alert, setAlert] = useState("");
  const {
    data: FetchData,
    error: ProductError,
    isLoading: loading,
  }: { data: any; error: any; isLoading: boolean } = useQuery(
    [`product/${id}`, id],
    () => SingleProductCall(id)
  );

  const { mutate: AddToWishlist, isLoading: WishListLoading } = useMutation(
    AddWishlistQuery,
    {
      onSuccess: (data: any) => {
        setAlert(data.msg);
      },
    }
  );

  const { isLoading: ChangingQty, mutate: PostQty } = useMutation(
    AddToCartQuery,
    {
      onSuccess: (data: any) => {
        queryClient.invalidateQueries(["Cart"]);
        data.msg && setAlert(data.msg); // if data.msg is not null
      },
      onSettled: () => {
        setTimeout(() => {
          setAlert("");
        }, 3000);
      },
      onError: (error: any) => {
        setAlert(error.msg);
      },
    }
  );

  const {
    data: WishListResult,
    isLoading: AddingWishList,
  }: {
    data: any;
    isLoading: boolean;
  } = useMutation(AddWishlistQuery, {
    onSuccess: (data: any) => {
      data.msg && setAlert(data.msg); // if data.msg is not null
    },
    onError: (error: any) => {
      setAlert(error.msg);
    },
  });
  const [Qty, setQty] = useState("1");

  const AddToCart = (e: React.ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();

    PostQty({
      id: Product._id,
      qty: parseInt(Qty),
    });
  };

  if (loading) return <Spinner />;

  if (ProductError) return <div className="Error">{ProductError}</div>;

  if (!FetchData) return null;

  const Product: DetailedProduct = FetchData;

  return (
    <div>
      <Helmet>
        <title>{Product.name}</title>
        <meta name="description" content={Product.description} />
      </Helmet>
      {FetchData && (
        <StyledContainer>
          {Alert && <AlertDisplay msg={Alert} type={true} />}
          <StyledProduct>
            <img src={Product.image} alt={Product.name} />
            <StyledDetails>
              <h2>{Product.name}</h2>
              <div className="star-review">
                <span className="stars">
                  <Rating
                    value={Product.rating}
                    text={`${Product.numReviews} reviews`}
                  />
                </span>
              </div>
              <h4>Price: {Product.price}</h4>
              <div className="Description">
                <div>Description:</div>
                {Product.description}
              </div>
            </StyledDetails>
            <StyledCheckout>
              <h3 className="status-label">Price: {Product.price}</h3>
              {Product.countInStock <= 0 ? (
                <h3 style={{ color: "red" }}>Status: Out In Stock</h3>
              ) : (
                <h3 className="status-label">Status: In Stock</h3>
              )}

              {Product.countInStock > 0 && (
                <form onSubmit={AddToCart}>
                  <label htmlFor="Qty">Qty</label>
                  <select id="Qty" onChange={(e) => setQty(e.target.value)}>
                    <Quantity countInStock={Product.countInStock} />
                  </select>
                  <br />

                  {localStorage.token ? (
                    <TimeoutBtn
                      Time={1000}
                      className="btn"
                      FormValue="ADD TO CART"
                    />
                  ) : (
                    <div className="btn-gray">Login/Register</div>
                  )}
                </form>
              )}
              <button
                className="btn btn-gray"
                onClick={() => {
                  AddToWishlist(Product._id);
                }}
              >
                WISHLIST
              </button>
              {ChangingQty && <AlertDisplay msg="Adding to cart" type={true} />}
              {WishListLoading && (
                <AlertDisplay msg="Adding to wishlist" type={true} />
              )}
              {AddingWishList && (
                <AlertDisplay msg={WishListResult.msg} type={true} />
              )}
            </StyledCheckout>
          </StyledProduct>
        </StyledContainer>
      )}
      <Reviews reviews={FetchData?.reviews} />
    </div>
  );
};

export default SingleProduct;
