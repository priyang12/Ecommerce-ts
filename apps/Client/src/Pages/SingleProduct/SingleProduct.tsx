import { useState } from "react";
import { useParams } from "react-router-dom";
import { DetailedProduct } from "../../interfaces";
import Reviews from "../../Components/Reviews";
import Rating from "../../Components/Rating";
import Quantity from "../../Components/Quantity";
import TimeoutBtn from "../../Components/TimeoutBtn";
import AlertDisplay from "../../Components/AlertDisplay";

//styles
import {
  StyledContainer,
  StyledProduct,
  StyledDetails,
  StyledCheckout,
} from "../../Components/StyledComponents/Productdetails";
import { AddOrUpdateCartQuery } from "../../API/CartAPI";
import { Helmet } from "react-helmet-async";
import { useSingleProduct } from "../../API/ProductAPI";
import { AddWishlistQuery } from "../../API/WishListAPI";
import Spinner from "../../Components/Spinner";

const SingleProduct = () => {
  const { id } = useParams<{ id: string }>();
  const [Alert, setAlert] = useState({
    msg: "",
    type: false,
  });

  const {
    data: FetchData,
    error: ProductError,
    isLoading: loading,
  } = useSingleProduct(id as string, false);

  const { mutate: AddToWishlist, isLoading: AddingWishList } =
    AddWishlistQuery(setAlert);

  const { isLoading: CartMutation, mutate: PostQty } =
    AddOrUpdateCartQuery(setAlert);

  const [Qty, setQty] = useState("1");

  const AddToCart = (e: React.ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    PostQty({
      ProductId: Product._id,
      qty: parseInt(Qty),
    });
  };

  if (loading) return <Spinner />;

  if (ProductError) return <div className="Error">{ProductError.message}</div>;

  if (!FetchData) return null;

  const Product: DetailedProduct = FetchData;

  return (
    <div>
      <Helmet>
        <title>{Product.name}</title>
        <meta name="description" content={Product.description} />
      </Helmet>
      {FetchData && (
        <StyledContainer>
          {Alert.msg && <AlertDisplay msg={Alert.msg} type={Alert.type} />}
          <StyledProduct>
            <img src={Product.image} alt={Product.name} />
            <StyledDetails>
              <h2>{Product.name}</h2>
              <div className="star-review">
                <span className="stars">
                  <Rating
                    value={Product.rating}
                    text={`${Product.numReviews} reviews`}
                  />
                </span>
              </div>
              <h4>Price: {Product.price}</h4>
              <div className="Description">
                <div>Description:</div>
                {Product.description}
              </div>
            </StyledDetails>
            <StyledCheckout>
              <h3 className="status-label">Price: {Product.price}</h3>
              {Product.countInStock <= 0 ? (
                <h3 style={{ color: "red" }}>Status: Out In Stock</h3>
              ) : (
                <h3 className="status-label">Status: In Stock</h3>
              )}

              {Product.countInStock > 0 && (
                <form onSubmit={AddToCart}>
                  <label htmlFor="Qty">Qty</label>
                  <select id="Qty" onChange={(e) => setQty(e.target.value)}>
                    <Quantity countInStock={Product.countInStock} />
                  </select>
                  <br />

                  {localStorage.token ? (
                    <TimeoutBtn
                      Time={1000}
                      className="btn"
                      FormValue="ADD TO CART"
                    />
                  ) : (
                    <div className="btn-gray">Login/Register</div>
                  )}
                </form>
              )}
              <button
                className="btn btn-gray"
                onClick={() => {
                  AddToWishlist(Product._id);
                }}
              >
                WISHLIST
              </button>
              {CartMutation && (
                <AlertDisplay msg="Adding to cart" type={true} />
              )}
              {AddingWishList && (
                <AlertDisplay msg="Adding to wishlist" type={true} />
              )}
            </StyledCheckout>
          </StyledProduct>
        </StyledContainer>
      )}
      <Reviews reviews={FetchData?.reviews} />
    </div>
  );
};

export default SingleProduct;
