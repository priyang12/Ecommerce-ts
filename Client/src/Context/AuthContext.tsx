import { AuthState, IContextModel } from './interfaces';
import { createContext, useReducer } from 'react';
import { AuthReducer } from './AuthReducer';

let Token = null;
if (localStorage.token) Token = localStorage.token;

const initialState: AuthState = {
  loading: false,
  err: null,
  token: Token,
  user: null,
};

export const AuthContext = createContext({} as IContextModel);

export const AuthProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(AuthReducer, initialState);
  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
