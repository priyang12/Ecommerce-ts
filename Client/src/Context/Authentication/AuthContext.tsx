import { AuthState, IContextModel } from "./interfaces";
import { createContext, useReducer } from "react";
import { AuthReducer } from "./AuthReducer";

let Token = null;
if (localStorage.token) Token = localStorage.token;

const user = sessionStorage.getItem("User");

const initialState: AuthState = {
  loading: true,
  err: null,
  token: Token,
  user: user && JSON.parse(user),
};

export const AuthContext = createContext({} as IContextModel);

export const AuthProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(AuthReducer, initialState);
  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
