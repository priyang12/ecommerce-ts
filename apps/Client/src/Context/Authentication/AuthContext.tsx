import React, { createContext, useReducer } from "react";
import { AuthActions, AuthReducer } from "./AuthReducer";
import { User } from "../../Types/interfaces";

export interface IContextModel {
  state: AuthState;
  dispatch: React.Dispatch<AuthActions>;
}
interface alert {
  show: boolean;
  message: string;
  type: string;
}

export type AuthUser = Omit<User, "password" | "updatedAt" | "createdAt">;

export interface AuthState {
  loading: boolean;
  err: string | null;
  token: string | null;
  user: AuthUser | null;
  alert: alert | null;
}

let Token = null;
if (localStorage.token) Token = localStorage.token;

const user = sessionStorage.getItem("User");

const initialState: AuthState = {
  loading: false,
  err: null,
  token: Token,
  user: user && JSON.parse(user),
  alert: null,
};

export const AuthContext = createContext({} as IContextModel);
AuthContext.displayName = "AuthContext";
export const useAuth = () => {
  return React.useContext(AuthContext);
};

export const AuthProvider = ({ children }: { children: any }) => {
  const [state, dispatch] = useReducer(AuthReducer, initialState);
  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
