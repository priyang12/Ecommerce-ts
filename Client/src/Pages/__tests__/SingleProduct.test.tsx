import { render, screen, waitFor } from "@testing-library/react";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { Router, Route } from "react-router-dom";
import { createMemoryHistory } from "history";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

//Component
import SingleProduct from "../SingleProduct";
import { AuthContext } from "../../Context/AuthContext";

const product = {
  rating: 4,
  numReviews: 1,
  price: 89.99,
  countInStock: 10,
  _id: "60d5e622e5179e2bb44bd838",
  name: "Airpods Wireless Bluetooth Headphones",
  image: "/Photos/image-1627384388351.webp",
  description:
    "Bluetooth technology lets you connect it with compatible devices wirelessly High-quality AAC audio offers immersive listening experience Built-in microphone allows you to take calls while working",
  brand: "Apple",
  category: "Electronics",
  user: "60d5e622e5179e2bb44bd835",
  reviews: [
    {
      _id: "6128b73ebc70bc35a0b0c3d0",
      name: "Priyang",
      rating: 4,
      comment: "very good",
      user: "6106f4c09d285d000436ed0a",
      createdAt: "2021-08-27T09:58:22.657Z",
      updatedAt: "2021-08-27T09:58:22.657Z",
    },
  ],
};

const mock = new MockAdapter(axios);

afterEach(() => {
  mock.reset();
});
const route = "/product/123123";
const history = createMemoryHistory({ initialEntries: [route] });

it("Mock Fetch Product Details", async () => {
  mock.onGet("/api/products/product/123123").reply(200, product);
  render(
    <Router history={history}>
      <Route path='/product/:id'>
        <SingleProduct />
      </Route>
    </Router>
  );
  await waitFor(() => {
    expect(
      screen.getByText(/Airpods Wireless Bluetooth Headphones/)
    ).toBeInTheDocument();
  });

  //Check for Review
  expect(screen.getByText(/REVIEWS/)).toBeInTheDocument();
});

it("Check SnapShot of Order", async () => {
  mock.onGet("/api/products/product/123123").reply(200, product);
  const { asFragment } = render(
    <Router history={history}>
      <Route path='/product/:id'>
        <SingleProduct />
      </Route>
    </Router>
  );
  await waitFor(() => {
    expect(
      screen.getByText(/Airpods Wireless Bluetooth Headphones/)
    ).toBeInTheDocument();
  });

  expect(asFragment()).toMatchSnapshot();
});

it("Mock Add to Cart", async () => {
  mock.onGet("/api/products/product/123123").reply(200, product);

  localStorage.setItem("token", "123123");

  render(
    <Router history={history}>
      <Route path='/product/:id'>
        <SingleProduct />
      </Route>
    </Router>
  );

  await waitFor(() => {
    expect(
      screen.getByText(/Airpods Wireless Bluetooth Headphones/)
    ).toBeInTheDocument();
  });
  const cartResponse = {
    msg: "Airpods Wireless Bluetooth Headphones is Added Cart",
  };
  mock.onPost("/api/cart").reply(200, cartResponse);
  userEvent.click(screen.getByText(/ADD TO CART/));

  await waitFor(() => screen.getByText(/Headphones is Added Cart/));
});
