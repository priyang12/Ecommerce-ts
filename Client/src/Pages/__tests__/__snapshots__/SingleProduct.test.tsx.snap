// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check SnapShot of Order 1`] = `
<DocumentFragment>
  <div>
    <section
      class="sc-hKwDye eScKcu"
    >
      <div
        class="sc-eCImPb fPFPeL"
      >
        <img
          alt="Airpods Wireless Bluetooth Headphones"
          src="/Photos/image-1627384388351.webp"
        />
        <div
          class="sc-jRQBWg iwhfcq"
        >
          <h2>
            Airpods Wireless Bluetooth Headphones
          </h2>
          <div
            class="star-review"
          >
            <span
              class="stars"
            >
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="far fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                1 reviews
              </span>
            </span>
          </div>
          <h4>
            Price: 89.99
          </h4>
          <div
            class="Description"
          >
            <div>
              Description:
            </div>
            Bluetooth technology lets you connect it with compatible devices wirelessly High-quality AAC audio offers immersive listening experience Built-in microphone allows you to take calls while working
          </div>
        </div>
        <div
          class="sc-gKclnd edbgla"
        >
          <h3
            class="status-label"
          >
            Price: 89.99
          </h3>
          <h3
            class="status-label"
          >
            Status: In Stock
          </h3>
          <form>
            <label>
              Qty
            </label>
            <select>
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
            </select>
            <br />
            <div
              class="btn-gray"
            >
              Login/Register
            </div>
          </form>
        </div>
      </div>
    </section>
    <div
      class="sc-bdvvtL qEhOG"
    >
      <h1>
        REVIEWS
      </h1>
      <ul
        class="sc-gsDKAQ gNizMB"
      >
        <li
          class="sc-dkPtRN ffMqkt"
        >
          <strong>
            Priyang
          </strong>
          <div
            class="star-review"
          >
            <span
              class="stars"
            >
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="fas fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span>
                <i
                  class="far fa-star"
                  style="color: rgb(248, 232, 37);"
                />
              </span>
              <span />
            </span>
          </div>
          <p>
            2021-08-27
          </p>
          <p>
            very good
          </p>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
